swarm/network/pss.go

```diff
@@ -1,52 +1,384 @@
+// - a collection of remote underlay address, mapped to the overlay addresses above
+// - a method to send a message to specific overlayaddr
 type Pss struct {
-	Overlay
-	LocalAddr	[]byte
-	C	chan []byte
+	Overlay                                              // we can get the overlayaddress from this
+	NodeId   *adapters.NodeId                            // we need the underlayaddr to make virtual p2p.Peers in the receiving end
+	PeerPool map[pot.Address]map[PssTopic]*PssReadWriter // keep track of all virtual p2p.Peers we are currently speaking to
+	//PeerReverse map[adapters.NodeId]pot.Address // keep track of all virtual p2p.Peers we are currently speaking to
+	handlers map[PssTopic]func([]byte, *p2p.Peer, []byte) error // topic and version based pss payload handlers
+	fwdcache map[pssDigest]time.Time                            // checksum of unique fields from pssmsg mapped to expiry, cache to determine whether to drop msg
+	cachettl time.Duration                                      // how long to keep messages in fwdcache
+	hasher   hash.Hash                                          // hasher to digest message to cache
 }
 
-func NewPss(k Overlay, addr []byte) *Pss {
+// 32bit checks should be enough for this use
+func (self *Pss) hashMsg(msg *PssMsg) pssDigest {
+	self.hasher.Reset()
+	self.hasher.Write(msg.To)
+	self.hasher.Write(msg.Payload.SenderUAddr)
+	self.hasher.Write(msg.Payload.SenderOAddr)
+	self.hasher.Write(msg.Payload.Topic[:])
+	self.hasher.Write(msg.Payload.Payload)
+	b := self.hasher.Sum([]byte{})
+	return pssDigest(binary.BigEndian.Uint32(b))
+}
+
+// todo error check overlay integrity
+func NewPss(k Overlay, cachettl time.Duration) *Pss {
+	if cachettl == 0 {
+		cachettl = defaultDigestCacheTTL
+	}
 	return &Pss{
-		Overlay: k,
-		LocalAddr: addr,
-		C: make(chan []byte),
+		Overlay:  k,
+		NodeId:   adapters.NewNodeId(k.GetAddr().UnderlayAddr()),
+		PeerPool: make(map[pot.Address]map[PssTopic]*PssReadWriter, PssPeerCapacity),
+		//PeerReverse: make(map[adapters.NodeId]pot.Address, PssPeerCapacity),
+		handlers: make(map[PssTopic]func([]byte, *p2p.Peer, []byte) error),
+		fwdcache: make(map[pssDigest]time.Time),
+		cachettl: cachettl,
+		hasher:   storage.MakeHashFunc("SHA3")(),
 	}
 }
 
-type PssMsg struct {
-	To   []byte
-	Data	[]byte
+// when full oldest out
+// periodically clean cache
+func (self *Pss) addToFwdCache(digest pssDigest) error {
+	self.fwdcache[digest] = time.Now().Add(self.cachettl)
+	return nil
 }
 
-func (pm *PssMsg) String() string {
-	return fmt.Sprintf("PssMsg: Recipient: %v", pm.To)
+// returns true if in cache AND not expired
+func (self *Pss) checkCache(digest pssDigest) bool {
+	entry, ok := self.fwdcache[digest]
+	if !ok || entry.Before(time.Now()) {
+		return false
+	}
+	return true
+}
+
+func (self *Pss) Register(name string, version int, handler func(msg []byte, p *p2p.Peer, from []byte) error) error {
```

Comment:
both Register and GetHandler needs to protect against concurrent calls with a lock